# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/Shape.cpp
# Commit: 11fc8f557a2a
# Full Hash: 11fc8f557a2a7cf3c89e7c1396316aabd9725778
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-03-06 09:33:16
# Regressor Bug: 1689413
# File Overlap Count: 1
# Description:
#   Bug 1689413 part 6 - Move Realm* from ObjectGroup to BaseShape. r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D106976
# ==============================================================================

diff -r 9bfa6023a553 -r 11fc8f557a2a js/src/vm/Shape.cpp
--- a/js/src/vm/Shape.cpp	Sat Mar 06 01:05:20 2021 +0000
+++ b/js/src/vm/Shape.cpp	Sat Mar 06 01:05:21 2021 +0000
@@ -329,8 +329,8 @@
   if (!shape->parent) {
     /* Treat as resetting the initial property of the shape hierarchy. */
     gc::AllocKind kind = gc::GetGCObjectKind(shape->numFixedSlots());
-    return EmptyShape::getInitialShape(cx, shape->getObjectClass(), proto, kind,
-                                       objectFlags);
+    return EmptyShape::getInitialShape(
+        cx, shape->getObjectClass(), shape->realm(), proto, kind, objectFlags);
   }
 
   Rooted<StackShape> child(cx, StackShape(shape));
@@ -1408,8 +1408,14 @@
 }
 
 inline BaseShape::BaseShape(const StackBaseShape& base)
-    : TenuredCellWithNonGCPointer(base.clasp) {
+    : TenuredCellWithNonGCPointer(base.clasp), realm_(base.realm) {
   MOZ_ASSERT(JS::StringIsASCII(clasp()->name));
+
+#ifdef DEBUG
+  if (GlobalObject* global = realm()->unsafeUnbarrieredMaybeGlobal()) {
+    AssertTargetIsNotGray(global);
+  }
+#endif
 }
 
 /* static */
@@ -1435,8 +1441,6 @@
   return nbase;
 }
 
-void BaseShape::traceChildren(JSTracer* trc) {}
-
 #ifdef DEBUG
 bool Shape::canSkipMarkingShapeCache() {
   // Check that every shape in the shape table will be marked by marking
@@ -1501,8 +1505,8 @@
     }
 
     using Lookup = InitialShapeEntry::Lookup;
-    Lookup lookup(shape->getObjectClass(), proto, shape->numFixedSlots(),
-                  shape->objectFlags());
+    Lookup lookup(shape->getObjectClass(), shape->realm(), proto,
+                  shape->numFixedSlots(), shape->objectFlags());
     InitialShapeSet::Ptr ptr = initialShapes().lookup(lookup);
     MOZ_RELEASE_ASSERT(ptr.found() && &*ptr == &r.front());
   }
@@ -1918,22 +1922,23 @@
 
 /* static */
 Shape* EmptyShape::getInitialShape(JSContext* cx, const JSClass* clasp,
-                                   TaggedProto proto, size_t nfixed,
-                                   ObjectFlags objectFlags) {
+                                   JS::Realm* realm, TaggedProto proto,
+                                   size_t nfixed, ObjectFlags objectFlags) {
+  MOZ_ASSERT(cx->compartment() == realm->compartment());
   MOZ_ASSERT_IF(proto.isObject(),
                 cx->isInsideCurrentCompartment(proto.toObject()));
 
-  auto& table = cx->zone()->initialShapes();
+  auto& table = realm->zone()->initialShapes();
 
   using Lookup = InitialShapeEntry::Lookup;
-  auto protoPointer =
-      MakeDependentAddPtr(cx, table, Lookup(clasp, proto, nfixed, objectFlags));
+  auto protoPointer = MakeDependentAddPtr(
+      cx, table, Lookup(clasp, realm, proto, nfixed, objectFlags));
   if (protoPointer) {
     return protoPointer->shape;
   }
 
   Rooted<TaggedProto> protoRoot(cx, proto);
-  StackBaseShape base(clasp);
+  StackBaseShape base(clasp, realm);
   Rooted<BaseShape*> nbase(cx, BaseShape::get(cx, base));
   if (!nbase) {
     return nullptr;
@@ -1944,7 +1949,7 @@
     return nullptr;
   }
 
-  Lookup lookup(clasp, protoRoot, nfixed, objectFlags);
+  Lookup lookup(clasp, realm, protoRoot, nfixed, objectFlags);
   if (!protoPointer.add(cx, table, lookup,
                         InitialShapeEntry(shape, protoRoot))) {
     return nullptr;
@@ -1955,9 +1960,10 @@
 
 /* static */
 Shape* EmptyShape::getInitialShape(JSContext* cx, const JSClass* clasp,
-                                   TaggedProto proto, gc::AllocKind kind,
+                                   JS::Realm* realm, TaggedProto proto,
+                                   gc::AllocKind kind,
                                    ObjectFlags objectFlags) {
-  return getInitialShape(cx, clasp, proto, GetGCKindSlots(kind, clasp),
+  return getInitialShape(cx, clasp, realm, proto, GetGCKindSlots(kind, clasp),
                          objectFlags);
 }
 
@@ -1986,7 +1992,7 @@
 void EmptyShape::insertInitialShape(JSContext* cx, HandleShape shape,
                                     HandleObject proto) {
   using Lookup = InitialShapeEntry::Lookup;
-  Lookup lookup(shape->getObjectClass(), TaggedProto(proto),
+  Lookup lookup(shape->getObjectClass(), shape->realm(), TaggedProto(proto),
                 shape->numFixedSlots(), shape->objectFlags());
 
   InitialShapeSet::Ptr p = cx->zone()->initialShapes().lookup(lookup);
@@ -2043,8 +2049,8 @@
     if (proto.isObject() && IsForwarded(proto.toObject())) {
       entry.proto = TaggedProto(Forwarded(proto.toObject()));
       using Lookup = InitialShapeEntry::Lookup;
-      Lookup relookup(shape->getObjectClass(), proto, shape->numFixedSlots(),
-                      shape->objectFlags());
+      Lookup relookup(shape->getObjectClass(), shape->realm(), proto,
+                      shape->numFixedSlots(), shape->objectFlags());
       e.rekeyFront(relookup, entry);
     }
   }